### DISCUSSION TITLE
When adding a dom and moving it on the same mutation, result on uninitialized alpine component

### DISCUSSION THREAD

**Discussion by Matsa59** [alpinejs/alpine#3947] [7. Bugs (archived)]
> Not related to phoenix LV but I'll explain it using LV

On phoenix LV, we have stream. When we add an element to the stream using `at` option (like push an element at the beginning of the stream), phoenix LV will perform 3 actions : `add`, `delete` and `add` again (we'll not talk about this behavior here)

The problem is Alpine have a code on mutation that tells: if an element is add and remove from mutation, don't initialized it.(https://github.com/alpinejs/alpine/blob/main/packages/alpinejs/src/mutation.js#L199)

It makes sens because we could consider it as a move operation but IMO we should ensure the component is NOT init instead of check if it's inside the remove list.

> https://github.com/alpinejs/alpine/discussions/1420 looks to be close to but not applicable anymore since the discover function is not available anymore

**Comment by SimoTod** [comment]
I don't think Alpine is able to tell if the component was already initialozed (the root node may or may not have an x-data attribute). If you have a hook you can use, you can call Alpine.initTree(el) where el is the element you are inserting in your page.

**Reply by Matsa59** [reply]
Unfortunately I don't have any hook on that part 

After some search I found a possible solution: if there is more added than remove it means the element is added. Otherwise it's a simple move

So the following piece of code could solve the issue

```js
function getNewNodes(addedNodes, removedNodes) {
  let remainingAddedNodes = [...addedNodes];

  removedNodes.forEach(removedNode => {
    let index = remainingAddedNodes.findIndex(addedNode => addedNode === removedNode)
    if (index > -1) {
      remainingAddedNodes.splice(index, 1)
    }
  });

  return remainingAddedNodes;
}
```

And so we could update the `mutate` function to be

```diff
+   const newNodes = getNewNodes(addedNodes, removedNodes)
+   for (let node of newNodes) {
-   for (let node of addedNodes) {
-       if (removedNodes.includes(node)) continue

        // If the node was eventually removed as part of one of his
        // parent mutations, skip it
        if (! node.isConnected) continue

        delete node._x_ignoreSelf
        delete node._x_ignore
        onElAddeds.forEach(i => i(node))
        node._x_ignore = true
        node._x_ignoreSelf = true
    }
```

**Reply by Matsa59** [reply]
A workaround for the moment is to add this piece of code to the app js

```js
export function onMutate(mutations) {
  let addedNodes = [];
  const moveCounter = new Map();

  for (let i = 0; i < mutations.length; i++) {
    const target = mutations[i].target;
    if (target._x_ignoreMutationObserver) continue;

    if (mutations[i].type === 'childList') {
      mutations[i].removedNodes.forEach(node => {
        if (node.nodeType === 1) {
          moveCounter.set(target, (moveCounter.get(target) || 0) - 1);
        }
      });
    }
  }

  for (let i = 0; i < mutations.length; i++) {
    const target = mutations[i].target;
    if (target._x_ignoreMutationObserver) continue;

    if (mutations[i].type === 'childList') {
      mutations[i].addedNodes.forEach(node => {
        if (node.nodeType === 1) {
          addedNodes.push({target, node});
          const count = moveCounter.get(target);
          if (count != null) {
            moveCounter.set(target, count + 1);
          }
        }
      });
    }
  }

  const movedNodes = new Set();

  for (let {node, target} of addedNodes) {
    if ((moveCounter.get(target) || 0) < 1) continue;

    movedNodes.add(node);
  }

  movedNodes.forEach((node) => {
    node._x_ignoreSelf = true;
    node._x_ignore = true;
  });

  for (let node of movedNodes) {
    // If the node was eventually removed as part of one of his
    // parent mutations, skip it
    if (! node.isConnected) continue;

    delete node._x_ignoreSelf;
    delete node._x_ignore;

    Alpine.initTree(node);
    node._x_ignore = true;
    node._x_ignoreSelf = true;
  }

  addedNodes.forEach((node) => {
    delete node._x_ignoreSelf;
    delete node._x_ignore;
  });
}

const observer = new MutationObserver(onMutate);
observer.observe(document, { subtree: true, childList: true, attributes: false });
```

> Note: this is needed only if someone use `at` option for phoenix LV streams

**Comment by ekwoka** [comment]
Alpine should probably do a better job of tracking if a tree is initialized or not....for many reasons.

**Reply by Matsa59** [reply]
I wonder if a simple `isInit` could be enough for the mutation system

because at the end, it will iterate over the `deletedNodes` to check if the node is moved or not. Trade some memory for a better CPU usage could be worth
